File: Java/DoublyC.java
// Doubly Circular Linked list 

class node
{
    public int data;
    public node next;  
    public node prev;  
}

class DoublyCLL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCLL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;

        if(first == null && last == null)
        {
            first = newn;
            last = newn;

            last.next = first;
            first.prev = last;
        }
        else
        {
           newn.next = first;
           first = newn;
           
           last.next = first;
           first.prev = last;
        }

        iCount++;
    }

    public void InsertLast(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;

        if(first == null && last == null)
        {
            first = newn;
            last = newn;

            last.next = first;
            first.prev = last;
        }
        else
        {
           last.next = newn;
           last = newn;

           last.next = first;
           first.prev = last;
        }

        iCount++;
        
    }
    
    public void InsertAtPos(int no, int pos)
    {
        node newn = null;
        node temp = null;
        int i = 0;

        if(pos < 1 || pos > iCount + 1)
        {
            System.out.println("Invalide Positon \n");
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount + 1)
        {
            InsertLast(no);
        }
        else 
        {
            newn = new node();
    
            newn.data = no;
            newn.next = null;
            newn.next = null;

            temp = first;

           for(i = 1; i < pos - 1; i++)
           {
              temp = temp.next;
           }

           newn.next = temp.next;
           temp.next.prev = newn;

           temp.next = newn;
           temp.prev = temp;

           iCount++;
        }  
    }
    
    public void DeleteFirst()
    {
       if(first == null && last == null)
       {
          return;
       }
       else if(first == last)
       {
          first = null;
          last = null;
       }
       else
       {
          first = first.next;

          last.next = first;
          first.prev = last;
       }

       iCount--;
    }

    public void DeleteLast()
    {
       node temp = null;

       if(first == null && last == null)
       {
          return;
       }
       else if(first == last)
       {
          first = null;
          last = null;
       }
       else
       {
          temp = first;

          while(temp.next != last)
          {
             temp = temp.next;
          }
          
          last = temp;

          last.next = first;
          first.prev = last;   
       }

       iCount--;
    }

    public void DeleteAtPos(int pos)
    {
        node temp = null;
        int iCnt = 0;

        if(pos < 1 || pos > iCount)
        {
            System.out.println("Invalide Positon \n");
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount)
        {
            DeleteLast();
        }
        else 
        {

           temp = first;

           for(iCnt = 1; iCnt < pos - 1; iCnt++)
           {
              temp = temp.next;
           }

           temp.next = temp.next.next;
           temp.next.prev = temp;

           iCount--;
        }

    }
    
    public void Display()
    {
        if(first == null && last == null)
        {
            return;
        }

        node temp = null;

        temp = first;

        do 
        { 
            System.out.print("| "+temp.data+" |<=> ");
            temp = temp.next;
        } while (temp != first);

        System.out.println();
    }

    public int Count()
    {
        return iCount;
    }
}

class DoublyC
{
    public static void main(String A[])
    {
        DoublyCLL sobj = new DoublyCLL();
        int iRet = 0;

        sobj.InsertFirst(41);
        sobj.InsertFirst(31);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(51);
        sobj.InsertLast(101);
        sobj.InsertLast(111);        
        sobj.InsertLast(121);

        sobj.Display();
    
        iRet = sobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        sobj.DeleteFirst();

        sobj.Display();
    
        iRet = sobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        sobj.DeleteLast();

        sobj.Display();
    
        iRet = sobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        sobj.InsertAtPos(105,5);

        sobj.Display();
    
        iRet = sobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        sobj.DeleteAtPos(5);

        sobj.Display();
    
        iRet = sobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);
    }
}

File: C/DoublyC.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

void InsertFirst(PPNODE first, PPNODE last, int no)
{
    PNODE newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if (*first == NULL)
    {
        *first = *last = newn;
        newn->next = newn;
        newn->prev = newn;
    }
    else
    {
        newn->next = *first;
        newn->prev = *last;

        (*first)->prev = newn;
        (*last)->next = newn;
        *first = newn;
    }
}

void InsertLast(PPNODE first, PPNODE last, int no)
{
    PNODE newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if (*first == NULL)
    {
        *first = *last = newn;
        newn->next = newn;
        newn->prev = newn;
    }
    else
    {
        newn->next = *first;
        newn->prev = *last;

        (*last)->next = newn;
        (*first)->prev = newn;
        *last = newn;
    }
}

int Count(PNODE first, PNODE last)
{
    if (first == NULL)
    {
        return 0;
    }

    int count = 0;
    PNODE temp = first;

    do
    {
        count++;
        temp = temp->next;
    } while (temp != first);

    return count;
}

void Display(PNODE first, PNODE last)
{
    if (first == NULL)
    {
        printf("List is empty\n");
        return;
    }

    PNODE temp = first;

    do
    {
        printf("| %d | <=> ", temp->data);
        temp = temp->next;
    } while (temp != first);
    printf("\n");
}

void InsertAtPos(PPNODE first, PPNODE last, int no, int pos)
{
    int size = Count(*first, *last);

    if (pos < 1 || pos > size + 1)
    {
        printf("Invalid position\n");
        return;
    }

    if (pos == 1)
    {
        InsertFirst(first, last, no);
    }
    else if (pos == size + 1)
    {
        InsertLast(first, last, no);
    }
    else
    {
        PNODE temp = *first;
        PNODE newn = (PNODE)malloc(sizeof(NODE));
        newn->data = no;

        for (int i = 1; i < pos - 1; i++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;
        newn->prev = temp;

        temp->next->prev = newn;
        temp->next = newn;
    }
}

void DeleteFirst(PPNODE first, PPNODE last)
{
    if (*first == NULL)
    {
        return;
    }
    else if (*first == *last)
    {
        free(*first);
        *first = *last = NULL;
    }
    else
    {
        PNODE temp = *first;

        *first = (*first)->next;
        (*first)->prev = *last;
        (*last)->next = *first;
        
        free(temp);
    }
}

void DeleteLast(PPNODE first, PPNODE last)
{
    if (*first == NULL)
    {
        return;
    }
    else if (*first == *last)
    {
        free(*first);
        *first = *last = NULL;
    }
    else
    {
        PNODE temp = *last;
        *last = (*last)->prev;
        (*last)->next = *first;
        (*first)->prev = *last;
        free(temp);
    }
}

void DeleteAtPos(PPNODE first, PPNODE last, int pos)
{
    int size = Count(*first, *last);

    if (pos < 1 || pos > size)
    {
        printf("Invalid position\n");
        return;
    }

    if (pos == 1)
    {
        DeleteFirst(first, last);
    }
    else if (pos == size)
    {
        DeleteLast(first, last);
    }
    else
    {
        PNODE temp = *first;

        for (int i = 1; i < pos - 1; i++)
        {
            temp = temp->next;
        }

        PNODE target = temp->next;

        temp->next = target->next;
        target->next->prev = temp;

        free(target);
    }
}


int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;

    InsertFirst(&head, &tail, 30);
    InsertFirst(&head, &tail, 20);
    InsertFirst(&head, &tail, 10);

    InsertLast(&head, &tail, 40);
    InsertLast(&head, &tail, 50);

    Display(head, tail);
    printf("Count: %d\n", Count(head, tail));

    InsertAtPos(&head, &tail, 25, 3);
    Display(head, tail);

    DeleteFirst(&head, &tail);
    Display(head, tail);

    DeleteLast(&head, &tail);
    Display(head, tail);

    DeleteAtPos(&head, &tail, 3);
    Display(head, tail);

    return 0;
}
